cheatsheets_draft/GraphQL_Cheat_Sheet.md

```diff
@@ -0,0 +1,257 @@
+    }
+}
+```
+
+***Infrastructure Timeout***
+
+Another option to add a timeout that is usually easier but less effective is adding a timeout on an HTTP server ([Apache/httpd](https://httpd.apache.org/docs/2.4/mod/core.html#timeout), [nginx](http://nginx.org/en/docs/http/ngx_http_core_module.html#send_timeout)), reverse proxy, or load balancer.
+
+#### Query Cost Analysis
+
+Query cost analysis involves assigning costs to the resolution of fields or types in incoming queries so that the server can reject queries that cost too much to run or will consume too many resources. This is not easy to implement and may not always be necessary but it is the most thorough approach to preventing DoS. See "Query Cost Analysis" in [this blog post](https://www.apollographql.com/blog/securing-your-graphql-api-from-malicious-queries-16130a324a6b) for more details on implementing this control.
+
+Apollo recommends:
+
+> **Before you go ahead and spend a ton of time implementing query cost analysis be certain you need it.** Try to crash or slow down your staging API with a nasty query and see how far you get — maybe your API doesn’t have these kinds of nested relationships, or maybe it can handle fetching thousands of records at a time perfectly fine and doesn’t need query cost analysis!
+
+APIs using graphql-java can utilize the built-in [MaxQueryComplexityInstrumentationto](https://github.com/graphql-java/graphql-java/blob/master/src/main/java/graphql/analysis/MaxQueryComplexityInstrumentation.java) to enforce max query complexity. APIs using JavaScript can utilize [graphql-cost-analysis](https://github.com/pa-bru/graphql-cost-analysis) or [graphql-validation-complexity](https://github.com/4Catalyzer/graphql-validation-complexity) to enforce max query cost.
+
+#### Rate Limiting
+
+Enforcing rate limiting on a per IP/user basis can help limit a single user's ability to spam requests to the service and impact performance. Ideally this can be done with a WAF, API gateway, or web server ([Nginx](https://www.nginx.com/blog/rate-limiting-nginx/), [Apache](https://httpd.apache.org/docs/2.4/mod/mod_ratelimit.html)/[HTTPD](https://github.com/jzdziarski/mod_evasive)) to reduce the effort of adding rate limiting.
+
+Or you could get somewhat complex with throttling and implement it in your code (non-trivial). See "Throttling" [here](https://www.howtographql.com/advanced/4-security/) for more about GraphQL-specific rate limiting.
+
+#### Server-side Batching and Caching
+
+To increase efficiency of a GraphQL API and reduce its resource consumption, [the batching and caching technique](https://graphql.org/learn/best-practices/#server-side-batching-caching) can be used to prevent making duplicate requests for pieces of data within a small time frame. Facebook's [DataLoader](https://github.com/facebook/dataloader) tool is one way to implement this.
+
+#### OS/Container Resource Management
+
+> Should we remove this section since it's not specific to GraphQL?
```

Comment:
Yeah, let's keep it for another cheat sheet? 😄

Reply:
Although I do agree this is not GraphQL specific, I think we should keep the recommendation. In the OWASP API Security Top 10 2019 this issue comes 4th.

You may argue that Query Limiting partially addresses the issue, but from my experience, quite often it is not done properly and in such cases Resources Limiting may prevent DoS.

Reply:
I do think this recommendation is important to have. Ideally this topic would be covered well in the [Web Service Security CS](https://cheatsheetseries.owasp.org/cheatsheets/Web_Service_Security_Cheat_Sheet.html#availability), since that can be thought of as a parent to this CS (and the REST Security CS). @PauloASilva maybe that would be a task for you being on the API security team? :) Or I could get a PR up to add something small to that CS. Then in this CS I can just link to that content.

I think I will remove the Docker specific items here since they are ripped straight out of the Docker Security CS. I will probably change that part to the below blurb instead:

> However, containerization platforms tend to make this task much easier. See the resource limiting section in the [Docker Security Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Docker_Security_Cheat_Sheet.html#rule-7-limit-resources-memory-cpu-file-descriptors-processes-restarts) for how to prevent DoS when using containers.

I will leave in the OS part about cgroups, ulimits, and LXC for now.

Reply:
> Ideally this topic would be covered well in the Web Service Security CS, since that can be thought of as a parent to this CS (and the REST Security CS).

I see your point. Unfortunately the Web Service Security CS is not as practice as this one, which I like most.
I'll open a PR to add a "Resources Limiting" section under "Availability".